working/macros/feature-specification.md

```diff
@@ -687,35 +687,134 @@ have the following restrictions:
+    APIs.
+
+1.  Invoke the corresponding visit method for all macros that implement phase 2
+    APIs.
+
+1.  Invoke the corresponding visit method for all macros that implement phase 3
+    APIs.
+
+While these are running, the macro will likely call back into the host
+environment to introspect over code in the current library cycle or previously
+compiled cycles. The introspection API is mostly syntactic and structural: a
+macro can walk the members on a class declaration or look at the *name* of a
+type annotation without the compiler having to do any resolution or type
+checking.
+
+When a macro wants to resolve an identifier in a type annotation, there is an
+explicit API for that. When that happens, the implementation attempts to resolve
+the identifier and return a reference to the resolved declaration. Macros do not
+have introspection access to the imperative code of a library, so that code
+doesn't need to be resolved or type-checked at this point.
+
+Meanwhile, the macro is also producing new declarations and definitions. These
+are collected and held by the macro processor. When introspecting over code, the
+implementation needs to show not just the state of the code on disk, but any of
+these new declarations produced previously by macros.
+
+#### 4. Generate an augmentation library
+
+Once all macro applications have finished running, the implementation creates a
+new empty augmentation library for each library containing macro applications.
+All of the declarations created by macros and held by the processor are now
+added to the augmentation.
+
+Entirely new declarations are simply added to the augmentation library as
+declarations. Declarations that wrap the original declaration's code are added
+as augmenting declarations. If a macro adds members to a type, then the type is
+added to the augmentation library as an augmenting type, and the members are
+added into that.
+
+**TODO**: How are name collisions from private declarations handled?
+
+The `Code` objects representing the signature and body of the declaration is
+serialized to Dart source. `Code` objects created from strings are inserted
+verbatim into the augmentation library. It's up to the macro author to take
+care when using unqualified identifiers in string-based `Code` objects.
+
+**TODO**: Do we want to find identifiers in string-based `Code` objects and
+implicitly scope them somehow?
+
+Instances of the `Identifier` class have special serialization. An import for
+library that the identifier resolves to is added to the augmentation with a new
+unique prefix identifier. The `Identifier` name is then serialized as a prefixed
+identifier for that prefix. (A more sophisticated implementation could reuse
+imports when multiple identifiers resolve to the same library, and may choose to
+omit the prefix entirely if the resulting identifier will still resolve
+correctly.)
+
+This augmentation library is written on disk in some implementation-defined
```

Comment:
I think that probably only the analyzer will actually write this to disk? Or possibly the CFE would in certain modes, but not others?

```suggestion
This augmentation library may be written on disk in some implementation-defined
```

Reply:
Yeah, good catch. Applied locally.